pkg/packetbrokeragent/translation.go

```diff
@@ -373,11 +373,17 @@ func toPBUplink(ctx context.Context, msg *ttnpb.GatewayUplinkMessage, config For
 				localization.Antennas = append(localization.Antennas, locAnt)
 			}
 
-			if md.Time != nil {
-				t := ttnpb.StdTime(md.Time)
-				if gatewayReceiveTime == nil || t.Before(*gatewayReceiveTime) {
-					gatewayReceiveTime = t
-				}
+			var t *time.Time
+			switch {
+			case md.GpsTime != nil:
```

Comment:
So this is what I requested in the issue, but maybe we need to do better - I think that if we have a GPS timestamp, we should override whatever we stored in `gatewayReceiveTime`, and no future updates to `gatewayReceiveTime` should be done.

Consider the following example:
2 metadata - one which contains a `gps_time` of `14:02:01.123123123` and one which contains a `time` of `14:02:01.0000000`. Since the secondary metadata is earlier we should use that, even though the GPS time is more precise and correct.

Reply:
But this does not solve the issue that I've mentioned - if the order of the metadata elements is 'low precision', 'high precision', 'low precision' you will keep the last timestamp, which is the low precision one.

Reply:
I see, so first choose most precise timestamp(s) and then choose among those in `gps` → `time` → `received_at` order?

Reply:
The axiom (which should be true in practice, but you never know) is that all `gps_time` is equivalent modulo 1 millisecond. So if you find a GPS time, then you don't need anything else. But if all of the metadata is missing the GPS time, then indeed we should take the earliest `time` and fallback to `received_at` (in case `time` is `nil`, compare using `received_at`).

Reply:
Solution proposal:

```go
gatewayReceiveTime = func() *time.Time {
	if t := md.GpsTime; t != nil {
		return ttnpb.StdTime(t)
	}

	t, r := ttnpb.StdTime(md.Time), ttnpb.StdTime(md.ReceivedAt)
	if t.IsZero() || r.Before(*t) {
		return r
	}
	return t
}()
```

Always use `gps_time` if available
```go
if t := md.GpsTime; t != nil {
  return ttnpb.StdTime(t)
}
```

Otherwise choose earliest from `time` and `received_at`.
Fallback to `received_at` if `time` is not available.
```go
t, r := ttnpb.StdTime(md.Time), ttnpb.StdTime(md.ReceivedAt)
if t.IsZero() || r.Before(*t) {
  return r
}
return t
```

Reply:
The problem is that when you do this, you do not take into consideration that `gatewayReceiveTime` could already be set to a `gps_time`.

Take the following:
Metadata 1 contains `gps_time`, metadata 2 contains `time`.
When you process metadata 1 you will just set `gatewayReceiveTime` to `gps_time`.
Then you process metadata 2, there is no `gps_time`, so you go into the secondary condition and overwrite the timestamp.

What I have in mind is:
```
switch {
case hasGPSTime:
case md.GpsTime != nil:
  gatewayReceiveTime = ttnpb.StdTime(md.GpsTime)
  hasGPSTime = true
case gatewayReceiveTime == nil || md.Time != nil && minimal condition between gatewayReceiveTime and md.Time:
  gatewayReceiveTime = ttnpb.StdTime(md.Time)
case gatewayReceiveTime == nill || minimal condition between gatewayReceiveTime and md.ReceivedAt:
  gatewayReceiveTime = ttnpb.StdTime(md.ReceivedAt)
}
```

Reply:
Pushed a new solution:

Add `var hasGpsTime bool` before the loop begins

Inside the loop:
```go
timestampCase := func(t *pbtypes.Timestamp) bool {
	g := gatewayReceiveTime
	return g == nil || t != nil && ttnpb.StdTime(t).Before(*g)
}
switch {
case hasGpsTime:
case md.GpsTime != nil:
	gatewayReceiveTime = ttnpb.StdTime(md.GpsTime)
	hasGpsTime = true
case timestampCase(md.Time):
	gatewayReceiveTime = ttnpb.StdTime(md.Time)
case timestampCase(md.ReceivedAt):
	gatewayReceiveTime = ttnpb.StdTime(md.ReceivedAt)
}
```